File: Array/src/main/python/solution15.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums = sorted(nums)
        print(nums)
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            res = self.get_two_nums(nums, i + 1, -nums[i])
            print(res)

            for l in res:
                l.insert(0, nums[i])

                ans.append(l)

        return ans

    def get_two_nums(self, nums, start, target):
        j = len(nums) - 1
        i = start
        ans = []

        while i < len(nums):
            if i > start and nums[i] == nums[i - 1]:
                i += 1
                continue
            while i < j and nums[i] + nums[j] > target:
                j -= 1

            if j > i and nums[i] + nums[j] == target:
                ans.append([nums[i], nums[j]])

            i += 1
        return ans


s = Solution()
res = s.threeSum([-1, 0, 1, 2, -1, -4])
print(res)

File: Array/src/main/java/double_pointer/Solution15.java
package double_pointer;

import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

public class Solution15 {
    public List<List<Integer>> threeSum(int[] nums) {

        Arrays.sort(nums);

        List<List<Integer>> ans = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            List<List<Integer>> twoNums = getTwoNums(-nums[i], nums, i + 1);

            if (twoNums.size() == 0) {
                continue;
            }
            // System.out.println(twoNums);
            for (List<Integer> l : twoNums) {
                ArrayList<Integer> innerAns = new ArrayList<>(l);
                innerAns.add(nums[i]);
                // System.out.println(innerAns);

                ans.add(innerAns);
            }

        }
        return ans;

    }

    private List<List<Integer>> getTwoNums(int target, int[] nums, int start) {
        List<List<Integer>> ans = new ArrayList<>();
        for (int i = start, j = nums.length - 1; i < j; i++) {
            // 需要i〉start； 如果〉0会把start 也有可能去掉
            if (i > start && nums[i] == nums[i - 1]) {
                continue;
            }

            // 需要 j>i; 否则j会多减一次，导致j=i
            while (j > i && nums[i] + nums[j] > target) {
                j--;
            }
            if (j > i && nums[i] + nums[j] == target) {
                ans.add(Arrays.asList(nums[i], nums[j]));

            }
        }
        return ans;

    }

    public static void main(String[] args) {
        int[] nums = { -1, 0, 1, 2, -1, -4 };
        // int[] nums = { 0, 1, 1 };
        // int[] nums = { 0, 0, 0, 0 };
        // int[] nums = { 1, 2, -2, -1 };
        Solution15 s = new Solution15();
        List<List<Integer>> res = s.threeSum(nums);

        System.out.println(res);
    }
}
